xskillscore/core/probabilistic.py

```diff
@@ -820,3 +822,244 @@ def _reliability(o, f, bin_edges):
+    fpr = xr.concat(fpr, "probability_bin")
+    tpr["probability_bin"] = bin_edges
+    fpr["probability_bin"] = bin_edges
+
+    fpr = fpr.fillna(1.0)
+    tpr = tpr.fillna(0.0)
+
+    # pad (0,0) and (1,1)
+    fpr_pad = xr.concat(
+        [
+            xr.ones_like(fpr.isel(probability_bin=0, drop=False)),
+            fpr,
+            xr.zeros_like(fpr.isel(probability_bin=-1, drop=False)),
+        ],
+        "probability_bin",
+    )
+    tpr_pad = xr.concat(
+        [
+            xr.ones_like(tpr.isel(probability_bin=0, drop=False)),
+            tpr,
+            xr.zeros_like(tpr.isel(probability_bin=-1, drop=False)),
+        ],
+        "probability_bin",
+    )
+
+    # https://github.com/scikit-learn/scikit-learn/blob/42aff4e2edd8e8887478f6ff1628f27de97be6a3/sklearn/metrics/_ranking.py#L916
+    # Attempt to drop thresholds corresponding to points in between and
+    # collinear with other points. These are always suboptimal and do not
+    # appear on a plotted ROC curve (and thus do not affect the AUC).
+    # Here xr.diff(_, 2) is used as a "second derivative" to tell if there
+    # is a corner at the point. (...) This keeps all cases where the point should be
+    # kept, but does not drop more complicated cases like fps = [1, 3, 7],
+    # tps = [1, 2, 4]; there is no harm in keeping too many thresholds.
+    if drop_intermediate and fpr.probability_bin.size > 2:
+
+        def _drop_intermediate(fpr, tpr):
+            optimal_idxs = xr.concat(
+                [
+                    fpr.isel(probability_bin=0, drop=False).astype("bool"),
+                    np.logical_or(
+                        fpr.diff("probability_bin", 2), tpr.diff("probability_bin", 2)
+                    ),
+                    fpr.isel(probability_bin=0, drop=False).astype("bool"),
+                ],
+                "probability_bin",
+            )
+            optimal_idxs["probability_bin"] = np.arange(
+                optimal_idxs.probability_bin.size
+            )
+            if isinstance(optimal_idxs, xr.Dataset):
+                optimal_idxs = optimal_idxs.to_array()
+            optimal_idxs = optimal_idxs.where(
+                optimal_idxs, drop=True
+            ).probability_bin.values
+            tpr = tpr.isel(probability_bin=optimal_idxs)
+            fpr = fpr.isel(probability_bin=optimal_idxs)
+            return fpr, tpr
+
+        fpr, tpr = _drop_intermediate(fpr, tpr)
+        fpr_pad, tpr_pad = _drop_intermediate(fpr_pad, tpr_pad)
+
+    def auc(fpr, tpr, dim="probability_bin"):
+        """Get area under the curve with trapez method."""
+        # reverse tpr, fpr to fpr, tpr, see numpy.trapz(y, x=None)
+        area = xr.apply_ufunc(
+            np.trapz, tpr, fpr, input_core_dims=[[dim], [dim]], dask="allowed"
+        )
+        if ((area > 1) | (area < 0)).any():
+            area = np.clip(area, 0, 1)  # allow only values between 0 and 1
+        return area
+
+    area = auc(fpr_pad, -tpr_pad)
```

Comment:
I'm still confused about why this is negative - don't we want the integral of the tpr with respect to the fpr?

Reply:
yes we do. but somehow I need this `-` to get the expected area.

You seems to handled this in doppyo with `abs`: https://github.com/csiro-dcfp/doppyo/blob/8c620889bde8a21d5937eb7ac71e72a040b31867/doppyo/skill.py#L773

maybe `np.abs()` is a cleaner solution? area should be positive anyways, better to understand than this `-`

Reply:
I think this makes sense to me as I think about it more. Increasing probability threshold (`bin_edges`) traces from _right to left_ on the ROC curve (see the figure below), so when `tpr` and `fpr` are sorted by `bin_edges`, `auc(fpr_pad, tpr_pad)` will be negative (because `dx` will be negative). I'm happy with either `np.abs` or `-`.
![Screen Shot 2021-01-29 at 10 39 56 am](https://user-images.githubusercontent.com/42455466/106212433-bc24c780-621e-11eb-876f-adcf61237812.png)